### ISSUE TITLE
[Notice] Major Mesa change

### ISSUE THREAD

**Issue by kallisti5** [haikuports/haikuports#8152]
This is a reminder of a major upcoming change to our OpenGL stack.
Going forward, we're moving to libglvnd to provide our OpenGL kit.  This is a big change and will require some coordination.

**Relevant upstream commits:**
* **Mesa** https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/21079 (merged)
* **libglvnd** https://gitlab.freedesktop.org/glvnd/libglvnd/-/merge_requests/279

**Summary:**
* libGL is now provided by libglvnd.  The recipe is complete and has been placed into haikuports as untested to prevent it building.
* mesa now provides the egl renderer which libglvnd uses.
  * These changes are not part of mesa 23.0.0, and will likely not show up until 23.1.0 or later (24.0.0?)
  * ~/config/non-packaged/add-ons/opengl/egl_vendor.d/...
    * 50_mesa.json  libEGL_mesa.so.0  libEGL_mesa.so.0.0.0  libglapi.so.0
```
cat 50_mesa.json 
{
    "file_format_version" : "1.0.0",
    "ICD" : {
        "library_path" : "./libEGL_mesa.so.0"
    }
}
```

tldr; Don't blindly update mesa!  There are likely going to be some regressions.  Rendering is working, but I noticed LaGrange (SDL based) is looking a bit odd.   The basic libglvnd unit tests are passing, but there are some failures we need to investigate.

**Comment by kallisti5** [comment]
Update: It looks like Mesa 23.1.0 got the Haiku changes for this one:
https://docs.mesa3d.org/relnotes/23.1.0.html

We should make a Mesa 23.1.1 recipe

**Comment by kallisti5** [comment]
EDIT: OLD. See latest comment [below](https://github.com/haikuports/haikuports/issues/8152#issuecomment-3702945059)


Ok.  I bumped the Mesa recipe to 23.1.1.

*Activation plan:*

We need to activate these at the same time.
  * Unblock mesa-23.1.1 on x86_64 (no longer provides libGL)
  * Unblock libglvnd-1.6.0 on x86_64 (now provides libGL)
  
We'll need vulkan, mesa, and libglvnd as build packages on modern gcc platforms.  Anything else? 

*Stable branches:*
* Do we need any patches on R1B4 to make it stable?

Testing notes:
  * GLTeapot - great
  * Haiku3D - great
  * GLInfo - ~crashes on close~ great after latest changes
  * Lagrange (SDL2) - regression from Mesa 22.x -  weird rendering issues, wobbly text
  * Supertuxkart (SDL2) - regression from Mesa 22.x - Screen flashing black / frame / black / frame
  * Nanosaur (SDL2) -  regression from Mesa 22.x  - Severe flashing black / frame / black / frame
  * Puri (Native) -  regression from Mesa 22.x - Changing to 3d chess mode.. crash on exit
  * GLTron (SDL1) - regression from Mesa 22.x - white screen. crash on exit
  * Quake 3 - TODO

*Status:*
I feel like we have too many regressions to activate these just yet.  The Screen flashing black / frame / black / frame feels like some front / back buffer issue.

**Comment by kallisti5** [comment]
@X547 ok.  I fixed the flashing in SDL apps! (ignore the DEBUG change)

```diffdiff --git a/src/egl/drivers/haiku/egl_haiku.cpp b/src/egl/drivers/haiku/egl_haiku.cpp
index 364bc5438f7..a4972ed58ea 100644
--- a/src/egl/drivers/haiku/egl_haiku.cpp
+++ b/src/egl/drivers/haiku/egl_haiku.cpp
@@ -68,7 +68,7 @@ extern "C" {
 #define BGL_SHARE_CONTEXT      1024
 
 
-#ifdef DEBUG
+#if 1
 #      define TRACE(x...) printf("egl_haiku: " x)
 #      define CALLED() TRACE("CALLED: %s\n", __PRETTY_FUNCTION__)
 #else
@@ -231,7 +231,7 @@ haiku_swap_buffers(_EGLDisplay *disp, _EGLSurface *surf)
        // flush back buffer and swap buffers if double buffering is used
        if (backBuffer != NULL) {
                screen->flush_frontbuffer(screen, st->pipe, backBuffer, 0, 0, buffer->winsysContext, NULL);
-               std::swap(frontBuffer, backBuffer);
+               //std::swap(frontBuffer, backBuffer);
                p_atomic_inc(&buffer->base.stamp);
        }
```

Any ideas here?

**Comment by X547** [comment]
Disabling swap is definitely not a solution. Proper rendering synchronization is needed.

**Comment by kallisti5** [comment]
@X547 do you have any ideas on what rendering synchronization is needed? 

To any outside observers:
* Frontbuffer is the rendering buffer
* Backbuffer is the rendering buffer used during Double Buffering (BGL_DOUBLE)

BGL_DOUBLE seems assumed now from libglvnd's Haiku driver.

**Comment by kenmays** [comment]
See: https://github.com/haikuports/haikuports/pull/7329
- Testing of Mesa 22.2.5

Add to current testing review:
- Colobot
- Speed Dreams

**Comment by X547** [comment]
> Puri (Native) - regression from Mesa 22.x - Changing to 3d chess mode.. crash on exit
> GLTron (SDL1) - regression from Mesa 22.x - white screen. crash on exit

https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30364

**Comment by X547** [comment]
Flickering and fail to switch between back and front buffer problems are fixed in https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30372.

**Comment by kenmays** [comment]
@X547  - Mesa 25.1.0-git (b14c8128) with libglvnd 1.7.0-git (606f6627) on hrev58792 x86 with Zink patches.
Builds successfully on Haiku x86. Mesa 25.1/libglvnd x86 packages installed successfully. I built your repo as well successfully.

Thanks for updating Zink driver. Will test moreso later.

**Comment by kallisti5** [comment]
How are we doing on this one?  I feel like "enough" of the bugs above have been addressed to do the mesa + libglvnd switch.


(Latest) Activation plan:

**Tasks TODO:**

  * We should likely update the mesa-24.1.4 recipe to 25.3.2
  * libglvnd-1.7.0 seems ready?

**We need to activate these at the same time.**

    Unblock mesa-25.3.2 on x86_64 (no longer provides libGL)
    Unblock libglvnd-1.7.0 on x86_64 (now provides libGL)

We'll need vulkan, mesa, and libglvnd as build packages on modern gcc platforms. Anything else?

**Stable branches:**

    Lets omit R1B5 plans, and target R1B6 (or whatever)

**Testing plan:**

Hopefully, the list of regressions with the mix above is way down from the [testing in 2023](https://github.com/haikuports/haikuports/issues/8152#issuecomment-1583484453). If yes... then we should be ready to move on this change.

- [ ] GLTeapot - TODO test
- [ ] Haiku3D - TODO test 
- [ ] GLInfo - [patch merged](https://review.haiku-os.org/c/haiku/+/7901) TODO Test
- [ ] Lagrange (SDL2) - [fixed in?](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30372) TODO test
- [ ] Supertuxkart (SDL2) - [fixed in?](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30372) TODO test
- [ ] Nanosaur (SDL2) - [fixed in?](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30372) TODO test
- [ ] Puri (Native) - [patch merged](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30364) TODO test
- [ ] GLTron (SDL1) -[patch merged](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30364) TODO test
- [ ] Quake 3 - TODO test


**X512's Zink / Nvidia hardware renderer**
Lets get upgraded to the new pipeline first, then we can work on merging the hardware rendering changes.

**Comment by waddlesplash** [comment]
We need (or at least really ought) to change the EGL ABI to support BViews instead of the "BitmapHook" system. I started looking at this, but right now Mesa 25 doesn't even build. With some minor changes I got it to the linking phase, but then it fails with lots of undefined references to `_mesa_glapi_tls_Context`. I haven't investigated that yet; I'll look at this more next week (if someone else doesn't beat me to it.)

**Comment by waddlesplash** [comment]
Also: we should build Gallium separate from the EGL drivers, so that we don't duplicate the rendering backends across GL/Vulkan and/or EGL Haiku + EGL Wayland. @X547 mentioned this as a possibility, I didn't look into what needs to be done to accomplish that yet (mostly just options passed to Mesa I suppose?)

**Comment by kallisti5** [comment]
https://glfs-book.github.io/glfs/shareddeps/mesa.html has a good (and recent) summary of the components.

Talking about how we want to divide the mesa package up:

**Things we need to figure out**
* mesa_eglvnd             - Do we want to keep this broken out?
* mesa_llvmpipe          - for the gallium  llvmpipe, llvm based rendering we have today (do we want to do away with this?)

**vulkan stuff**
* mesa_zink                  - for the gallium  OpenGL -> Vulkan pipeline X512 has been working on
* mesa_vlk_llvmpipe    - aka, vulkan-drivers=swrast, aka "lvp" and lavapipe (replace mesa_llvmpipe?)
* mesa_vlk_nouveau    - for the Vulkan-based renderer X512 has been working on.
* mesa_vlk_xyz             - for future Vulkan-based renderers

**Comment by kenmays** [comment]
> We should likely update the mesa-24.1.4 recipe to 25.3.2

- Mesa 24.3.4 builds fine on Haiku R1B5 x64. Tested on R1B5 hrev59261 x64 and x86.
- Mesa 25.3.3 builds - but stops at the newer vulkanlite renderdoc implementation. Needs review.

**Comment by kallisti5** [comment]
@kenmays both 25.3.3 bugs fixed in upstream mesa.    I'm working on a 25.3.3 package now.

**Comment by kallisti5** [comment]
mesa-25.3.3-1 + libglvnd 1.7.0-4 + hrev59261

- [X] GLTeapot - Great
- [X] Haiku3D - Great
- [X] GLInfo - Great
- [X] Neverball 1.6.1 Great
- [ ] Neverputt - Crashes sometimes
- [ ] BillardGL - Crashes glutDestroyWindow Segment violation
- [ ] Lagrange (SDL2) - [fixed in?](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30372) same issues
- [ ] Supertuxkart (SDL2) - [fixed in?](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30372) TODO test
- [X] Nanosaur (SDL2) - [fixed in?](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30372) Ok. Crashes on exit.. maybe not GL related
- [ ] Puri (Native) - [patch merged](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30364) 3d mode crashes on click.  ChessBoardView::MouseDown(BPoint
- [ ] GLTron (SDL1) -[patch merged](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/30364) white screen, crashes on exit. UnlockGL called from wrong thread. see https://dev.haiku-os.org/ticket/6400
- [ ] Quake 3 - TODO test

**Comment by Peppersawce** [comment]
> Quake 3 - TODO test

I'm noticing this test has been a TODO for a bit, I assume most people testing this are on 64 bit and Quake 3 is not available there (sort of, FTEQW can run it), ioquake3 has (had?) the UnlockGL issue when compiled on 64 bit as well.

At this point OpenArena might be a more convenient test target, it's in the Depot and it's the same engine as Quake 3, updates and patches aside.

As a starting point:

Openarena (Mesa 22.0.5) - the window flashes white on startup (intro video issues?) then it works ok, crashes on exit.

**Comment by kenmays** [comment]
Confirmed mesa 25.3.3 build on hrev59264 x64.
```
~/haikuports/sys-libs/mesa> ls -al ~/mesa_25.3.3
total 51710
drwxr-xr-x 1 user root     2048 Jan  3 02:18 .
drwxr-xr-x 1 user root    40960 Jan  3 02:18 ..
-rw-r--r-- 1 user root 45427881 Jan  3 02:18 mesa_debuginfo-25.3.3-1-x86_64.hpkg
-rw-r--r-- 1 user root     5508 Jan  3 02:18 mesa_devel-25.3.3-1-x86_64.hpkg
-rw-r--r-- 1 user root  4622427 Jan  3 02:18 mesa_eglvnd-25.3.3-1-x86_64.hpkg
-rw-r--r-- 1 user root  2827383 Jan  3 02:18 mesa_lavapipe-25.3.3-1-x86_64.hpkg
-rw-r--r-- 1 user root      851 Jan  3 02:18 mesa-25.3.3-1-x86_64.hpkg
```

Note: I'll only focus on the x64 (64-bit) platform testing for now.

**Comment by kenmays** [comment]
Mesa 25.3.3 (llvmpipe) w/libglvnd 1.7.0-git (606f6627) on Haiku R1B5 (hrev59273) x64: 
- 0AD - gaming works fine
- AAAA (Adamant Armor Affection Adventure) -gaming works OK
- Antigravitaattor -gaming works OK
- Armagetron Advanced - gaming works fine
- Astromenace - gaming works fine
- Better Spades - white windowed screen during launch
- Blender3 - worked fine
- BillardGL - gaming works fine
- BTanks - white windowed screen during launch
- BZFlag - gaming works fine
- ClassiCube - gaming works fine
- Colobot - Missing textures and player
- Critical Mass - gameplay works fine (Needs libGL.so symlinked)
- Dust Racer 2D - black screen on launch (can see hand cursor). Qt5.
- Eternal Lands - gaming works fine
- Extreme Tux Racer - gaming works fine
- FooBillard++ - gaming works fine
- GLtron - gaming works fine
- Goonies - white windowed screen during launch
- Kobo Deluxe - gaming works fine
- Kubrick - game crashes on launch
- Lagrange - worked fine
- Minetest - gaming works fine
- Nanosaur - gaming works fine
- Neverball - gaming works fine
- Neverputt - gaming works fine
- Openarena - white screen during launch
- OpenClonk - gaming works fine
- Simutrans - gaming works fine
- Speed Dreams 2 - gaming works fine
- Stellarium - crashed on launch
- SuperTuxKart - gaming works fine
- Teeworld - gaming works fine
- Widelands -gaming works fine
- Xmoto - white windowed screen during launch

Screensavers:
- GLife - white screen ( no crashes)

**Comment by X547** [comment]
> crashed on launch

Can you provide stack trace of crash?

**Comment by kenmays** [comment]
> Can you provide stack trace of crash?

@X547 - Yes. I'll put error message and crash reports here.

**Comment by Peppersawce** [comment]
> Openarena (Mesa 22.0.5) - the window flashes white on startup (intro video issues?) then it works ok, crashes on exit.

Openarena (Mesa 25.3.3) - Only displays a white window, doesn't crash on exit anymore.

**Comment by kenmays** [comment]
Mesa 25.3.3 testing:

> Openarena
`Only displays a white window, doesn't crash on exit anymore.`

> **GLife** (core Haiku screensaver)
```
----- Initializing Renderer ----
-------------------------------
----- R_Init -----
Cmd_AddCommand: paletteglsldump already defined
Initializing OpenGL display
...setting mode -1: 1281 1025
Using 8/8/8 Color bits, 24 depth, 8 stencil display.
Available modes: '1280x1024 640x480 800x600 1024x768 1920x1080'
GL_RENDERER: llvmpipe (LLVM 21.1.8, 256 bits)
Initializing OpenGL extensions
...ignoring GL_EXT_texture_compression_s3tc
...ignoring GL_S3_s3tc
...using GL_EXT_texture_env_add
...using GL_ARB_multitexture
...using GL_EXT_compiled_vertex_array
...ignoring GL_EXT_texture_filter_anisotropic
...ignoring GL_ARB_vertex_shader
...GL_EXT_paletted_texture not found
Initializing Shaders
    
GL_VENDOR: Mesa
GL_RENDERER: llvmpipe (LLVM 21.1.8, 256 bits)
GL_VERSION: 4.5 (Compatibility Profile) Mesa 25.3.3 (git-aa72a69664)
...
PIXELFORMAT: color(24-bits) Z(24-bit) stencil(8-bits)
MODE: -1, 1280 x 1024 fullscreen hz:N/A
GAMMA: software w/ 0 overbright bits
rendering primitives: single glDrawElements
texturemode: GL_LINEAR_MIPMAP_NEAREST
picmip: 1
texture bits: 0
multitexture: enabled
compiled vertex arrays: enabled
texenv add: enabled
compressed textures: disabled
glsl programs: disabled
----- finished R_Init -----
Loading vm file vm/ui.qvm...
File "vm/ui.qvm" found in "/boot/system/apps/OpenArena/baseoa/pak6-patch088.pk3"
...which has vmMagic VM_MAGIC_VER2
Loading 1555 jump table targets
VM file ui compiled to 785004 bytes of code
ui loaded in 1848448 bytes on the hunk
58 arenas parsed
33 bots parsed
tty]Kill Thread
``
```

**Comment by kenmays** [comment]
@kallisti5 - llvmpipe, zink, and NVK.

Good strategy overall.

So far, the new graphics stack is doing very well for the majority of the modern 3D games, demos, and applications I've tested from Haikuports as of Haiku R1B5 (hrev59273).

**Comment by kenmays** [comment]
LockGL/UnlockGL behavior seems different for specific apps (Qt-based, SDL-based, certain screensavers).

libglvnd recipe:
Fix libGL.so->libGL.so.1 symlink structure. Example:
```
~/Desktop> ls /system/lib/libGL*
/system/lib/libGL.so
/system/lib/libGL.so.1
/system/lib/libGL.so.1.0.0
```

Btanks 0.9.8083-4:
```
haiku_create_window_surface
UnlockGL is called from wrong thread, lockerThread: 3695, callerThread: 3711
Unlocking BLocker with sem 2657556 from wrong thread 3711, current holder 3695 (see issue #6400).
[20:29:19.350][engine/src/config.cpp:296]        [debug] cleaning up config...
Kill Thread
```

Colobot: Gold Edition 0.2.2-alpha
```
[INFO]: Colobot: Gold Edition 0.2.2-alpha starting
Mesa: error: GL_INVALID_OPERATION in glBlitFramebuffer(mismatched samples)
```

Stellarium 25.4-1:
```
Debug information for team /boot/system/apps/Stellarium (2357):
Haiku revision: hrev59273 Jan  3 2026 07:02:33 (x86_64)

Active Threads:
	thread 2360: BApplication_thread 
	thread 2372: team 2357 debug task 
	thread 2357: Stellarium (main)
		state: Call (thread 2357 tried accessing address 0x824633cff0 which is not allocated (base: 0x824633cff0, size: 16, alignment: 16, allocated by thread: 2357, freed by thread: 2357))

		Frame		IP			Function Name
		-----------------------------------------------
		00000000	0x2141f8d8bd7	_kern_debugger + 0x7 
			Disassembly:
				_kern_debugger:
				0x000002141f8d8bd0:   48c7c0ed000000  mov $0xed, %rax
				0x000002141f8d8bd7:             0f05  syscall <--

		0x7fe52fda41d0	0x2141f9655ed	panic(char const*, ...) + 0xad 
		0x7fe52fda4230	0x2141f9659e9	guarded_heap_segfault_handler(int, __siginfo_t*, void*) + 0x149 
		0x7fe52fda4240	0x7feada73c23c	commpage_signal_handler + 0x2c 
		0x7fe52fda47b0	0x295524270a	haiku_initialize(_egl_display*) + 0xa 
		0x7fe52fda47f0	0x1deabaa3709	BGLView::Display::Init() + 0x229 
		0x7fe52fda4820	0x1deabaa3a35	BGLView::Display::FirstReferenceAcquired() + 0x25 
		0x7fe52fda4840	0x3665f12860	BReferenceable::AcquireReference() + 0x50 
		0x7fe52fda4870	0x1deabaa3a88	BGLView::Renderer::Init() + 0x18 
		0x7fe52fda48b0	0x1deabaa41f7	BGLView::BGLView(BRect, char*, unsigned long, unsigned long, unsigned long) + 0xe7 
		0x7fe52fda4950	0xda21e6420e	OSMesaCreateContextExt + 0x19e 
		0x7fe52fda4a20	0xda21e6512e	QHaikuGLContext::QHaikuGLContext(QOpenGLContext*) + 0x1ce 
		0x7fe52fda4a40	0xda21e65d43	non-virtual thunk to QHaikuIntegration::createPlatformOpenGLContext(QOpenGLContext*) const + 0x23 
		0x7fe52fda4a60	0x20aef69bbe	QOpenGLContext::create() + 0x2e 
		0x7fe52fda4bc0	0x20aef8e98f	QRhiGles2::create(QFlags<QRhi::Flag>) + 0x132f 
		0x7fe52fda4c30	0x20aee12eb6	QRhi::create(QRhi::Implementation, QRhiInitParams*, QFlags<QRhi::Flag>, QRhiNativeHandles*) + 0xa6 
		0x7fe52fda4cd0	0x20aec61731	QBackingStoreRhiSupport::create() + 0x1d1 
		0x7fe52fda4d70	0x174d3e00596	QWidgetPrivate::create() + 0x4c6 
		0x7fe52fda4dd0	0x174d3e0086d	QWidget::create(unsigned long long, bool, bool) + 0xdd 
		0x7fe52fda4e30	0x174d3e09f4f	QWidget::setWindowState(QFlags<Qt::WindowState>) + 0x14f 
		0x7fe52fda4e50	0x174d3e0a564	QWidget::showFullScreen() + 0x24 
		0x7fe52fda4ee0	0x16582dba6eb	StelMainView::setFullScreen(bool) + 0x1b 
		0x7fe52fda5380	0x16582b07f12	main + 0x20f2 
		0x7fe52fda53b0	0x16582b1b1c9	_start + 0x39 
		0x7fe52fda53e0	0x5ae604f3d3	runtime_loader + 0x113 
		00000000	0x7feada73c258	commpage_thread_exit + 0 
```

3D-Pipes (screensaver):
```
Debug information for team /boot/system/preferences/ScreenSaver (4191):
Haiku revision: hrev59273 Jan  3 2026 07:02:33 (x86_64)

Active Threads:
	thread 4191: ScreenSaver (main)
	thread 4197: w>Password Window 
	thread 4209: llvmpipe-0 
	thread 4210: llvmpipe-1 
	thread 4211: llvmpipe-2 
	thread 4212: llvmpipe-3 
	thread 4213: pthread func 
	thread 4214: pthread func 
	thread 4215: pthread func 
	thread 4216: pthread func 
	thread 4217: ÃSÒÿÿÿ:disk$0 
	thread 4218: ScreenSaverRenderer 
	thread 4219: team 4191 debug task 
	thread 4198: w>ScreenSaver 
		state: Call (generic segfault)

		Frame		IP			Function Name
		-----------------------------------------------
		00000000	0x1157d651bd7	_kern_debugger + 0x7 
			Disassembly:
				_kern_debugger:
				0x000001157d651bd0:   48c7c0ed000000  mov $0xed, %rax
				0x000001157d651bd7:             0f05  syscall <--

		0x7fac88679540	0x1157d6de5ed	panic(char const*, ...) + 0xad 
		0x7fac88679550	0x7ff40f38123c	commpage_signal_handler + 0x2c 
		0x7fac88679af0	0x1ebf749e91b	llvm::TimeTraceProfiler::begin(std::basic_string<char, std::char_traits<char>, std::allocator<char> >, llvm::function_ref<std::basic_string<char, std::char_traits<char>, std::allocator<char> > ()()>, llvm::TimeTraceEventType) + 0xab 
		0x7fac88679b60	0x1ebf74a287b	llvm::timeTraceProfilerBegin(llvm::StringRef, llvm::StringRef) + 0xcb 
		0x7fac88679bb0	0x1ebfc357ebc	llvm::TimeProfilingPassesHandler::runBeforePass(llvm::StringRef, llvm::Any) + 0x9c 
		0x7fac88679c00	0x1ebfc3580d7	_ZN4llvm6detail18UniqueFunctionBaseIvJNS_9StringRefENS_3AnyEEE8CallImplIZNS_26TimeProfilingPassesHandler17registerCallbacksERNS_28PassInstrumentationCallbacksEEUlS2_S3_E_EEvPvS2_RS3_ + 0x97 
		0x7fac88679d20	0x1ebf7737a19	_ZN4llvm11PassManagerINS_6ModuleENS_15AnalysisManagerIS1_JEEEJEE3runERS1_RS3_ + 0x429 
		0x7fac8867b120	0x1ebfc339cc3	runPasses(llvm::Module*, llvm::Function*, char const*, llvm::TargetMachine*, llvm::LLVMPassBuilderOptions*) + 0xfd3 
		0x7fac8867b560	0x1942557c803	lp_passmgr_run + 0x63 
		0x7fac8867b6b0	0x1942551b6a1	gallivm_compile_module + 0x1a1 
		0x7fac8867ba10	0x19425614c7e	generate_setup_variant + 0xfbe 
		0x7fac8867ba60	0x194256152b0	llvmpipe_update_setup + 0x350 
		0x7fac8867bad0	0x194255f945b	llvmpipe_update_derived + 0x13b 
		0x7fac8867bb30	0x194255d9093	llvmpipe_draw_vbo + 0x3a3 
		0x7fac8867bb80	0x19424ead645	st_draw_gallium_multimode + 0x65 
		0x7fac8867bc10	0x19424f29068	vbo_exec_vtx_flush + 0x538 
		0x7fac8867bc30	0x19424f279cb	vbo_exec_FlushVertices + 0x14b 
		0x7fac8867bc60	0x1942501f460	pop_matrix + 0xc0 
		0x7fac8867bc80	0x1942501fe83	_mesa_PopMatrix + 0x33 
		0x7fac8867bcf0	0x10abe9b9d2b	PipesGLView::DrawPipes() + 0x8b 
		0x7fac8867bd10	0x10abe9ba6b0	PipesGLView::Draw(BRect) [clone .localalias] + 0x60 
		0x7fac8867bd40	0x24ee29ef7f	BView::_Draw(BRect) + 0x9f 
		0x7fac8867bf70	0x24ee2ac807	BWindow::DispatchMessage(BMessage*, BHandler*) + 0xcb7 
		0x7fac8867c000	0x24ee2a7d73	BWindow::task_looper() + 0x1d3 
		0x7fac8867c020	0x24ee1e858b	BLooper::_task0_(void*) + 0x1b 
		0x7fac8867c040	0x1157d6508c7	thread_entry + 0x17 
		00000000	0x7ff40f381258	commpage_thread_exit + 0 
```

**Comment by Peppersawce** [comment]
I've spotted another regression with Mesa 25. I have the impression some sort of texture/bitmap scaling is enabled and it is messing up some 2D-based games.
<img width="1075" height="807" alt="Image" src="https://github.com/user-attachments/assets/32fa8aec-1d4e-4a50-8efd-4e36aa3dc117" />
It's very apparent in OpenRCT2, the logo, the tracks and the in-game text (among other things) all suffer from this. Moving the options pop-up around does change how the text is rendered in it.

To note: text written using (I think) actual fonts looks fine (lower-left corner) but most of the in-game text uses the original game's "font", which is very likely a collection of bitmap tiles.

EDIT: The note is incorrect, those pieces of text can also become distorted when changing resolutions or moving the window around.

Changing the resolution also makes the top horizontal line of the window tweak out.

Of course, all these elements rendered fine with Mesa 22.

> Mesa 22.0.5 (llvmpipe) w/libglvnd-current:

To be fair, I'm pretty sure Mesa 22 cannot use libglvnd so that would explain the output you got with OpenArena.

**Comment by kenmays** [comment]
> To be fair, I'm pretty sure Mesa 22 cannot use libglvnd so that would explain the output you got.

This was to track Mesa | libGLvnd interaction for debugging. I think I found most of the overall GL issues and segmentation violations (guarded_heap) using llvmpipe.

**Comment by kallisti5** [comment]
@kenmays to be clear, you're testing:

libglvnd current + mesa 25.3.3 recipe with mesa_llvmpipe?

To be clear, there really haven't been any changes between libglvnd current and 1.7.0. [ref](https://gitlab.freedesktop.org/glvnd/libglvnd/-/commits/master)

**Comment by kenmays** [comment]
> libglvnd current + mesa 25.3.3 recipe with mesa_llvmpipe?

Yes. All mesa packages are v25.3.3. using libglvnd-current. Mesa-lllvmpipe driver (software rendering only)..
Platform: Haiku R1B5 (hrev59275) x64

<img width="559" height="385" alt="Image" src="https://github.com/user-attachments/assets/d964fb00-9a4d-4666-bbcf-5050fe24b5c3" />

**Comment by kenmays** [comment]
- Tested all currently ported 3D software in Haiku/Haikuports up to Haiku R1B6-dev (hrev60102) and Mesa 26.2.2 on x86-64 platform. 
-  See @x512 and @kallisti5 for further resolutions.

**Comment by lonemadmax** [comment]
While testing for some video artifacts with a new KDE app I enabled libglvnd and mesa 25, and the app crashed.

In `egl_haiku.cpp::haiku_display_destroy` we free the haiku_egl_display, but the _EGLDisplay keeps a pointer to it in its DriverData field. The next initialization with the same EGLDisplay finds it already has a haiku_egl_display (pointing to freed memory), increments its ref count (unrelated to the issue: shouldn't this be using atomics?) and happily goes with it until something crashes. The Stellarium crash above is very similar to the ones I get.

Adding `disp->DriverData = NULL;` fixes the crash. Not my KDE/Qt6 issues though, I get a black window with this version.

**Comment by kenmays** [comment]
Packages: libglvnd 1.7.0 / Mesa 26.2.2
Tested on: Haiku R1B6-dev (hrev60080) x64

1. All Haiku 3D demos work.
2. GLife 3D screensaver still not working. All of the other core 3D screensavers are working. 

SOLVED: Glife uses the old HGL pipeline of Mesa 22.0.5 and needs tweaking for the newer "EGL" pipeline in Mesa 26.2.2. I got it working now through the "new" 3D pipeline and even did some review on the older 3dfx Voodoo variant of GLife. This clears all current Haiku distributed 3D screensavers/demos provided on the LiveDVD. (big smile)

Certain Haikuport's 3D screensaver will need tweaks for the new pipeline r maybe provide a fallback for previous HGL setup - BDirectWindow versus BAttachWindow, Lock() versus LockGL behavior, etc, etc..., see: Gravity versus GLife context setup.
 
<img width="1131" height="795" alt="Image" src="https://github.com/user-attachments/assets/1e1cb372-e697-4336-b5f5-36d6ec934be5" />
